test/functional/api/buffer_updates_spec.lua

```diff
@@ -0,0 +1,737 @@
+  helpers.command, helpers.eval, helpers.nvim, helpers.next_msg
+
+local origlines = {"original line 1",
+                   "original line 2",
+                   "original line 3",
+                   "original line 4",
+                   "original line 5",
+                   "original line 6"}
+
+local function expectn(name, args)
+  -- expect the next message to be the specified notification event
+  eq({'notification', name, args}, next_msg())
+end
+
+local function sendkeys(keys)
+  nvim('input', keys)
+  -- give neovim some time to process msgpack requests before possibly sending
+  -- more key presses - otherwise they all pile up in the queue and get
+  -- processed at once
+  local ntime = os.clock() + 0.1
+  repeat until os.clock() > ntime
+end
+
+local function open(activate, lines)
+  local filename = helpers.tmpname()
+  helpers.write_file(filename, table.concat(lines, "\n").."\n", true)
+  command('edit ' .. filename)
+  local b = nvim('get_current_buf')
+  -- what is the value of b:changedtick?
+  local tick = eval('b:changedtick')
+
+  -- turn on live updates, ensure that the nvim_buf_updates_start messages
+  -- arrive as expectected
+  if activate then
+    ok(buffer('updates_start', b, true))
+    expectn('nvim_buf_updates_start', {b, tick, lines, false})
+  end
+
+  return b, tick, filename
+end
+
+local function editoriginal(activate, lines)
+  if not lines then
+    lines = origlines
+  end
+  -- load up the file with the correct contents
+  helpers.clear()
+  return open(activate, lines)
+end
+
+local function reopen(buf, expectedlines)
+  ok(buffer('updates_stop', buf))
+  expectn('nvim_buf_updates_end', {buf})
+  -- for some reason the :edit! increments tick by 2
+  command('edit!')
+  local tick = eval('b:changedtick')
+  ok(buffer('updates_start', buf, true))
+  expectn('nvim_buf_updates_start', {buf, tick, expectedlines, false})
+  command('normal! gg')
+  return tick
+end
+
+local function reopenwithfolds(b)
+  -- discard any changes to the buffer
+  local tick = reopen(b, origlines)
+
+  -- use markers for folds, make all folds open by default
+  command('setlocal foldmethod=marker foldlevel=20')
+
+  -- add a fold
+  command('2,4fold')
+  tick = tick + 1
+  expectn('nvim_buf_update', {b, tick, 1, 4, {'original line 2/*{{{*/',
+                                          'original line 3',
+                                          'original line 4/*}}}*/'}})
+  -- make a new fold that wraps lines 1-6
+  command('1,6fold')
+  tick = tick + 1
+  expectn('nvim_buf_update', {b, tick, 0, 6, {'original line 1/*{{{*/',
+                                          'original line 2/*{{{*/',
+                                          'original line 3',
+                                          'original line 4/*}}}*/',
+                                          'original line 5',
+                                          'original line 6/*}}}*/'}})
+  return tick
+end
+
+describe('liveupdate', function()
```

Comment:
describe('buffer events', function()

and "knows" can be removed from the descriptions below.